### PR TITLE
Loading precomputed Gramian (kernel) matrix from file 

### PR DESCRIPTION
**Pull request by mpecha**
The feature also contains implementation of new interface function SVMComputeOperator() and related binary implementation SVMComputeOperator_Binary(). Currently, the function support only building the Hessian matrix (related to underlying SVM-QP) implicitly. In future, the function will contains calling of functions related to building non-linear kernel matrix or explicit Hessian. Other implementations and/or reimplementations are related to the concept of this feature.

### CODE REVIEW

File: examples/ex1.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  char           test_file[PETSC_MAX_PATH_LEN]     = "data/heart_scale.t.bin";
+  PetscBool      test_file_set = PETSC_FALSE;
+
+  PetscViewer    viewer;
+  PetscErrorCode ierr;
+
+  ierr = PermonInitialize(&argc,&argv,(char *)0,help); if (ierr) return ierr;
+
+  ierr = PetscOptionsGetString(NULL,NULL,"-f_training",training_file,sizeof(training_file),NULL);CHKERRQ(ierr);
+  ierr = PetscOptionsGetString(NULL,NULL,"-f_kernel",kernel_file,sizeof(test_file),&test_file_set);CHKERRQ(ierr);
```
Comment (jkruzik) [inline examples/ex1.c]:
```suggestion
  ierr = PetscOptionsGetString(NULL,NULL,"-f_kernel",kernel_file,sizeof(kernel_file),NULL);CHKERRQ(ierr);
```
otherwise setting kernel without setting test_file will load default test_file and run tests with it

File: src/svm/impls/binary/binary.c [hunk 2]
```diff
@@ -491,142 +566,229 @@ PetscErrorCode SVMUpdate_Binary_Private(SVM svm)
...
   if (svm->posttraincalled) {
     TRY( SVMUpdate_Binary_Private(svm) );
     PetscFunctionReturn(0);
   }
 
-  if (svm->hyperoptset) {
-    TRY( SVMGridSearch(svm) );
-  }
+  /* TODO generalize implementation of hyper parameter optimization */
+  if (svm->hyperoptset) { TRY( SVMGridSearch(svm) ); }
```
Comment (jkruzik) [inline src/svm/impls/binary/binary.c]: Wrong code style (Developers guide 2.2.1 13,14)
```suggestion
  if (svm->hyperoptset) {
    TRY( SVMGridSearch(svm) );
  }
```

File: src/svm/impls/binary/binary.c [hunk 3]
```diff
@@ -698,26 +858,25 @@ PetscErrorCode SVMSetUp_Binary(SVM svm)
...
     TRY( QPSSetFromOptions(qps) );
   }
-
   TRY( QPSSetUp(qps) );
 
-  /* decreasing reference counts */
-  TRY( MatDestroy(&X_training) );
-  TRY( MatDestroy(&H) );
+  /* Create work vectors */
+  for (i = 0; i < 3; ++i) { TRY( VecDestroy(&svm_binary->work[i]) ); }
```
Comment (jkruzik) [inline src/svm/impls/binary/binary.c]:
```suggestion
  for (i = 0; i < 3; ++i) {
    TRY( VecDestroy(&svm_binary->work[i]) );
  }
```

File: src/svm/impls/binary/binary.c [hunk 4]
```diff
@@ -1549,6 +1708,73 @@ PetscErrorCode SVMGridSearch_Binary(SVM svm)
...
+PetscErrorCode SVMLoadMatGramian_Binary(SVM svm,PetscViewer v)
+{
+  MPI_Comm comm;
+  Mat      G;
+
+  PetscFunctionBegin;
+  TRY( PetscObjectGetComm((PetscObject) svm,&comm) );
+
+  /* Create matrix */
+  TRY( MatCreate(comm,&G) );
```
Comment (jkruzik) [inline src/svm/impls/binary/binary.c]: use
```
PetscObjectComm((PetscObject) svm)
```
when comm is not argument to several functions

File: src/svm/impls/binary/binary.c [hunk 5]
```diff
@@ -1549,6 +1708,73 @@ PetscErrorCode SVMGridSearch_Binary(SVM svm)
...
+  TRY( MatSetType(G,MATDENSE) );
+  TRY( PetscObjectSetName((PetscObject) G,"G") );
+  TRY( PetscObjectSetOptionsPrefix((PetscObject) G,"gramian_") );
```
Comment (jkruzik) [inline src/svm/impls/binary/binary.c]: inherit prefix from the svm object?

File: src/svm/impls/binary/binary.c [hunk 6]
```diff
@@ -1549,6 +1708,73 @@ PetscErrorCode SVMGridSearch_Binary(SVM svm)
...
+
+  Mat        G;
+  PetscInt   M,N;
+
+  PetscBool  isascii;
+
+  PetscFunctionBegin;
+  TRY( SVMGetMatGramian(svm,&G) );
+  if (!G) {
+    TRY( PetscObjectGetComm((PetscObject) v,&comm) );
```
Comment (jkruzik) [inline src/svm/impls/binary/binary.c]: use
```
PetscObjectComm((PetscObject) v)
```
in SETERRQ unless comm is available from previous code

File: src/svm/impls/binary/binary.c [hunk 7]
```diff
@@ -1549,6 +1708,73 @@ PetscErrorCode SVMGridSearch_Binary(SVM svm)
...
+    TRY( PetscViewerASCIIPushTab(v) );
+    TRY( PetscViewerASCIIPrintf(v,"dimensions: %D,%D\n",M,N) );
+    TRY( PetscViewerASCIIPopTab(v) );
+
+    TRY(PetscViewerASCIIPopTab(v));
+  } else {
+    TRY( PetscObjectGetComm((PetscObject) v,&comm) );
+    TRY( PetscObjectGetType((PetscObject) v,&type_name) );
+
+    FLLOP_SETERRQ1(comm,PETSC_ERR_SUP,"Viewer type %s not supported for SVMViewMatGramian",type_name);
```
Comment (jkruzik) [inline src/svm/impls/binary/binary.c]:
```
PetscObjectComm((PetscObject) v)
```

File: src/svm/interface/svm.c [hunk 8]
```diff
@@ -1511,6 +1511,80 @@ PetscErrorCode SVMViewScore(SVM svm,PetscViewer v)
...
+.seealso SVM, SVMSetMatGramian()
+@*/
+PetscErrorCode SVMGetMatGramian(SVM svm,Mat *G)
+{
+
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(svm,SVM_CLASSID,1);
+  PetscValidPointer(G,2);
+
+  TRY( PetscTryMethod(svm,"SVMGetMatGramian_C",(SVM,Mat *),(svm,G)) );
```
Comment (jkruzik) [inline src/svm/interface/svm.c]: Use vs Try. Does the user always expect to get G?
https://lists.mcs.anl.gov/pipermail/petsc-dev/2016-April/018920.html

File: src/svm/interface/svm.c [hunk 9]
```diff
@@ -2257,6 +2335,78 @@ PetscErrorCode SVMComputeModelParams(SVM svm)
...
+  const char *dataset_file;
+
+  PetscFunctionBeginI;
+  PetscValidHeaderSpecific(svm,SVM_CLASSID,1);
+  PetscValidHeaderSpecific(v,PETSC_VIEWER_CLASSID,2);
+
+  PetscLogEventBegin(SVM_LoadGramian,svm,0,0,0);
+  if (svm->ops->loadgramian) {
+    TRY( svm->ops->loadgramian(svm,v) );
+  }
```
Comment (jkruzik) [inline src/svm/interface/svm.c:2450]: error out when implementation is not available?

File: src/svm/interface/svm.c [hunk 10]
```diff
@@ -2257,6 +2335,78 @@ PetscErrorCode SVMComputeModelParams(SVM svm)
...
+  PetscLogEventEnd(SVM_LoadGramian,svm,0,0,0);
+
+  TRY( PetscViewerFileGetName(v,&dataset_file) );
+  TRY( PetscStrcpy(svm->kernel_mat_file,dataset_file) );
```
Comment (jkruzik) [inline src/svm/interface/svm.c]: Use
```
TRY( PetscViewerFileGetName(v,&svm->kernel_mat_file) );
```

File: src/svm/impls/binary/binary.c [hunk 11]
```diff
@@ -1549,6 +1708,73 @@ PetscErrorCode SVMGridSearch_Binary(SVM svm)
...
+    TRY( PetscObjectGetComm((PetscObject) v,&comm) );
+    FLLOP_SETERRQ(comm,PETSC_ERR_ARG_NULL,"Gramian (kernel) matrix is not specified");
+  }
+
+  TRY( PetscObjectTypeCompare((PetscObject)v,PETSCVIEWERASCII,&isascii) );
+  if (isascii) {
+    TRY( PetscObjectPrintClassNamePrefixType((PetscObject) svm,v) );
+
+    TRY( PetscViewerASCIIPushTab(v) );
+    TRY( PetscViewerASCIIPrintf(v,"Gramian (kernel) matrix was loaded from file \"%s\" successfully!\n",svm->kernel_mat_file) );
```
Comment (jkruzik) [inline src/svm/impls/binary/binary.c]: We should throw an error when the file load failed. Then we do not need this line and can get rid of svm->kernel_mat_file.

File: examples/ex1.c [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+/*TEST
+
+  test:
+    filter: grep -v MPI
+    args: -qps_view_convergence -svm_view -svm_view_score
+    args: -f_training $PERMON_SVM_DIR/examples/data/heart_scale.bin -f_kernel $PERMON_SVM_DIR/examples/data/heart_scale.kernel.bin
+    args: -f_test $PERMON_SVM_DIR/examples/data/heart_scale.t.bin
+    output_file: output/exbinfile_1.out
+TEST*/
```
Comment (jkruzik) [inline examples/ex1.c]: new line at the end of file

File: examples/ex1.c [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+  ierr = SVMCreate(PETSC_COMM_WORLD,&svm);CHKERRQ(ierr);
+  ierr = SVMSetType(svm,SVM_BINARY);CHKERRQ(ierr);
+  ierr = SVMSetFromOptions(svm);CHKERRQ(ierr);
+
+  /* Load training dataset */
+  ierr = PetscViewerBinaryOpen(PETSC_COMM_WORLD,training_file,FILE_MODE_READ,&viewer);CHKERRQ(ierr);
+  ierr = SVMLoadTrainingDataset(svm,viewer);CHKERRQ(ierr);
+  ierr = PetscViewerDestroy(&viewer);CHKERRQ(ierr);
+
+  /* Load Gramian matrix */
```
Comment (jkruzik) [inline examples/ex1.c:38]: consider adding
```
if (kernel_file_set) {
```
then this would be the simplest example for all binary type SVM (no complications with different file viewers).

File: src/svm/impls/binary/binary.c (hunk 5 above)
Comment (mpecha) [reply src/svm/impls/binary/binary.c]: Basically, it's a good idea when we think about complex machine learning application containing thousands of components, e.g. classifiers, dictionaries, autoencoders, data analysis part etc. I think current development stage of the package is not ready for implementing in such complex application because some necessities and standard functionalities missing like multiclass/multilabel approaches, implicit/explicit kernels, one-class classification, regressions (logistic, isotonic), v-SVM, lp-SVM, etc. Moreover, applying this idea should also be done in other parts of package, e.g. loading training/test dataset. Therefore, I've decided to not implement it in this PR.

File: src/svm/interface/svm.c (hunk 9 above)
Comment (mpecha) [reply src/svm/interface/svm.c:2450]: Great! I think it should be useful to implement an error message in this case. However, I followed the PETSc and PERMON(QP) implementations of similar functions, and I didn't see it. I support this idea, however, not in this PR. I'll create an issue associated with your idea.

File: examples/ex1.c (hunk 1 above)
Comment (mpecha) [reply examples/ex1.c]: Accepted.

File: src/svm/impls/binary/binary.c (hunk 2 above)
Comment (mpecha) [reply src/svm/impls/binary/binary.c]: Accepted.

File: src/svm/impls/binary/binary.c (hunk 3 above)
Comment (mpecha) [reply src/svm/impls/binary/binary.c]: Accepted.

File: src/svm/impls/binary/binary.c (hunk 4 above)
Comment (mpecha) [reply src/svm/impls/binary/binary.c]: Accepted.

File: src/svm/impls/binary/binary.c (hunk 6 above)
Comment (mpecha) [reply src/svm/impls/binary/binary.c]: Accepted.

File: src/svm/impls/binary/binary.c (hunk 7 above)
Comment (mpecha) [reply src/svm/impls/binary/binary.c]: Accepted.

File: src/svm/interface/svm.c (hunk 8 above)
Comment (mpecha) [reply src/svm/interface/svm.c]: Okay, my fault. When the G is not available, it does not make sense to expect it. Fixed.

File: src/svm/interface/svm.c (hunk 10 above)
Comment (mpecha) [reply src/svm/interface/svm.c]: Regarding to face-to-face discussion, I've deciced to remove name of file.

File: src/svm/impls/binary/binary.c (hunk 11 above)
Comment (mpecha) [reply src/svm/impls/binary/binary.c]: No, inner implementation of PETSc throws an error automatically. So, such code is pointless.

File: examples/ex1.c (hunk 12 above)
Comment (mpecha) [reply examples/ex1.c]: Accepted.

File: examples/ex1.c (hunk 13 above)
Comment (mpecha) [reply examples/ex1.c:38]: Accepted.

File: examples/ex1.c [hunk 14]
```diff
@@ -0,0 +1,70 @@
+
+static char help[] = "Trains binary SVM classification model using precomputed Gramian (linear kernel) matrix.\n\
+                      Provided Gramian matrix is computed and related to no-bias SVM formulation (svm_binary_mod 2).\n";
```
Comment (jkruzik) [inline examples/ex1.c]: Remove mention of the precomputed Gramian (it is not neccessary to set it). Remove second line (any kernel file can be used). Describe input parameters as in e.g.,  examples/exbinfile.c.

Comment (mpecha) [reply examples/ex1.c]: Good point! It makes sense to write the example more generally. I agree. Accepted.

**Review by jkruzik — APPROVED** [review]

File: src/svm/interface/svm.c [hunk 15]
```diff
@@ -2257,6 +2335,73 @@ PetscErrorCode SVMComputeModelParams(SVM svm)
...
+
+  Input Parameters:
++ svm - SVM context
+- v - viewer
+
+ Level: intermediate
+
+.seealso SVM, SVMViewMatGramian(), SVMLoadTrainingDataset()
+@*/
+PetscErrorCode SVMLoadMatGramian(SVM svm,PetscViewer v)
```
Comment (haplav) [inline src/svm/interface/svm.c]: - [x] You have `SVMViewGramian` but `SVMLoadMatGramian`. I think you should remove `Mat` from the name. The same for Get, Set above.
- [ ] Why is this needed? Why not just call `MatLoad` and `SVMSetGramian`? Why this needs to be implementation-specific?

File: src/svm/interface/svm.c [hunk 16]
```diff
@@ -2257,6 +2335,73 @@ PetscErrorCode SVMComputeModelParams(SVM svm)
...
+    TRY( PetscObjectGetComm((PetscObject) v,&comm) );
+    TRY( SVMViewMatGramian(svm,PETSC_VIEWER_STDOUT_(comm)) );
+  }
+  PetscFunctionReturnI(0);
+}
+
+#undef __FUNCT__
+#define __FUNCT__ "SVMViewGramian"
+/*@
+  SVMViewGramian - Visualizes (precomputed) Gramian matrix.
```
Comment (haplav) [inline src/svm/interface/svm.c:2465]: - [ ] Why is this needed? Why not just call `SVMGetGramian` and `MatView`? Why this needs to be implementation-specific?
- [x] Manpage speaks about `SVMViewGramian` but the function is actually named `SVMViewMatGramian`. I would omit that `Mat`.

File: src/svm/interface/svm.c [hunk 17]
```diff
@@ -1511,6 +1511,80 @@ PetscErrorCode SVMViewScore(SVM svm,PetscViewer v)
...
+  Collective on SVM
+
+  Level: advanced
+
+  Input parameter:
+. svm - SVM context
+
+.seealso SVM, SVMSetMatGramian()
+@*/
+PetscErrorCode SVMComputeOperator(SVM svm)
```
Comment (haplav) [inline src/svm/interface/svm.c]: This function couple two things together:
1. computes the operator
2. sets it as a Hessian of the underlying QP

It could just return the operator and it would then be set separately using SVMSetOperator. I don't like such coupling and doing things under the hood.

**Review by haplav — CHANGES_REQUESTED** [review]

File: src/svm/interface/svm.c (hunk 17 above)
Comment (mpecha) [reply src/svm/interface/svm.c]: Okay, I agree. I thought it was early to generalize this thing. However, I've changed my mind when I saw your suggestion. Accepted.

File: src/svm/interface/svm.c (hunk 15 above)
Comment (mpecha) [reply src/svm/interface/svm.c]: Okay, it will be removed. Regarding the second one, this interface function is designed for using the package user-friendly.

File: src/svm/interface/svm.c (hunk 16 above)
Comment (mpecha) [reply src/svm/interface/svm.c:2465]: The reason is same as above - for using the package user-friendly. Regading omiting  ``Mat`` from func. name, I argee.

File: src/svm/impls/binary/binary.c [hunk 18]
```diff
@@ -871,11 +923,12 @@ PetscErrorCode SVMSetUp_Binary(SVM svm)
   for (i = 0; i < 3; ++i) {
     TRY( VecDestroy(&svm_binary->work[i]) );
   }
-  TRY( MatCreateVecs(Xt,NULL,&svm_binary->work[0]) );
+  TRY( MatCreateVecs(Xt,NULL,&svm_binary->work[0]) ); /* TODO use duplicated vector y instead of creating vec? */
```
Comment (jkruzik) [inline src/svm/impls/binary/binary.c:926]: this is fine

File: src/svm/impls/binary/binary.c [hunk 19]
```diff
@@ -943,7 +996,7 @@ PetscErrorCode SVMReconstructHyperplane_Binary(SVM svm)
 {
   SVM_Binary *svm_binary = (SVM_Binary *) svm->data;
 
-  QPS       qps;
+  /* QPS       qps; */
```
Comment (jkruzik) [inline src/svm/impls/binary/binary.c]: remove

**Review by jkruzik — APPROVED** [review]

File: src/svm/impls/binary/binary.c (hunk 19 above)
Comment (mpecha) [reply src/svm/impls/binary/binary.c]: Thanks! I forgot about this.
